This is an odd failure: https://github.com/ocaml/ocaml/actions/runs/6729852164/job/18291457670?pr=12710

```
the file './ocamlopt' has not the right magic number: expected Caml1999X033, got 
make[2]: *** [Makefile:2324: ocamldoc/odoc_types.cmx] Error 127
make[2]: *** Waiting for unfinished jobs....
make[3]: Leaving directory '/home/runner/work/ocaml/ocaml'
```

I'll restart this job.

--- a/runtime/caml/major_gc.h
+++ b/runtime/caml/major_gc.h
@@ -45,13 +45,8 @@ void caml_finish_major_cycle(int force_compaction);
 #ifdef DEBUG
 int caml_mark_stack_is_empty(void);
 #endif
-/* Ephemerons and finalisers */
-void caml_orphan_allocated_words(void);
-void caml_add_to_orphaned_ephe_list(struct caml_ephe_info* ephe_info);
-void caml_add_orphaned_finalisers (struct caml_final_info*);
-void caml_final_domain_terminate (caml_domain_state *domain_state);
-void caml_incr_num_domains_orphaning_finalisers (void);
-void caml_decr_num_domains_orphaning_finalisers (void);
+void caml_orphan_ephemerons(caml_domain_state*);
+void caml_orphan_finalisers(caml_domain_state*);
 
 /* Forces finalisation of all heap-allocated values,
    disregarding both local and global roots.

--- a/runtime/caml/major_gc.h
+++ b/runtime/caml/major_gc.h
@@ -50,6 +50,8 @@ void caml_orphan_allocated_words(void);
 void caml_add_to_orphaned_ephe_list(struct caml_ephe_info* ephe_info);
 void caml_add_orphaned_finalisers (struct caml_final_info*);
 void caml_final_domain_terminate (caml_domain_state *domain_state);
+void caml_incr_num_domains_orphaning_finalisers (void);
+void caml_decr_num_domains_orphaning_finalisers (void);
 
 /* Forces finalisation of all heap-allocated values,
    disregarding both local and global roots.

--- a/runtime/domain.c
+++ b/runtime/domain.c
@@ -1817,40 +1817,6 @@ static void caml_atfork_default(void)
 
 CAMLexport void (*caml_atfork_hook)(void) = caml_atfork_default;
 
-static void orphan_ephemerons(caml_domain_state* domain_state)
-{
-  if (domain_state->ephe_info->todo == 0 &&
-      domain_state->ephe_info->live == 0 &&
-      domain_state->ephe_info->must_sweep_ephe == 0)
-    return;
-
-  caml_add_to_orphaned_ephe_list(domain_state->ephe_info);
-  CAMLassert (domain_state->ephe_info->live == 0);
-  CAMLassert (domain_state->ephe_info->todo == 0);
-}
-
-static void orphan_finalisers(caml_domain_state* domain_state)
-{
-  struct caml_final_info* f = domain_state->final_info;
-
-  if (f->todo_head != NULL || f->first.size != 0 || f->last.size != 0) {
-    /* have some final structures */
-    if (caml_gc_phase != Phase_sweep_and_mark_main) {
-      /* Force a major GC cycle to simplify constraints for orphaning
-         finalisers. See note attached to the declaration of
-         [num_domains_orphaning_finalisers] variable in major_gc.c */
-      caml_incr_num_domains_orphaning_finalisers();
-      caml_finish_major_cycle(0);
-      caml_decr_num_domains_orphaning_finalisers();
-      CAMLassert(caml_gc_phase == Phase_sweep_and_mark_main);
-    }
-    caml_add_orphaned_finalisers (f);
-    /* Create a dummy final info */
-    domain_state->final_info = caml_alloc_final_info();
-  }
-  caml_final_domain_terminate(domain_state);
-}
-
 static inline int domain_terminating(dom_internal *d) {
   return d->interruptor.terminating;
 }
@@ -1888,8 +1854,9 @@ static void domain_terminate (void)
        STW sections that has sent an interrupt to this domain. */
 
     caml_finish_marking();
-    orphan_ephemerons(domain_state);
-    orphan_finalisers(domain_state);
+
+    caml_orphan_ephemerons(domain_state);
+    caml_orphan_finalisers(domain_state);
 
     /* take the all_domains_lock to try and exit the STW participant set
        without racing with a STW section being triggered */

--- a/runtime/domain.c
+++ b/runtime/domain.c
@@ -1836,9 +1836,12 @@ static void handover_finalisers(caml_domain_state* domain_state)
   if (f->todo_head != NULL || f->first.size != 0 || f->last.size != 0) {
     /* have some final structures */
     if (caml_gc_phase != Phase_sweep_and_mark_main) {
-      /* Force a major GC cycle to simplify constraints for
-       * handing over finalisers. */
+      /* Force a major GC cycle to simplify constraints for orphaning
+         finalisers. See note attached to the declaration of
+         [num_domains_orphaning_finalisers] variable in major_gc.c */
+      caml_incr_num_domains_orphaning_finalisers();
       caml_finish_major_cycle(0);
+      caml_decr_num_domains_orphaning_finalisers();
       CAMLassert(caml_gc_phase == Phase_sweep_and_mark_main);
     }
     caml_add_orphaned_finalisers (f);

--- a/runtime/domain.c
+++ b/runtime/domain.c
@@ -1817,7 +1817,7 @@ static void caml_atfork_default(void)
 
 CAMLexport void (*caml_atfork_hook)(void) = caml_atfork_default;
 
-static void handover_ephemerons(caml_domain_state* domain_state)
+static void orphan_ephemerons(caml_domain_state* domain_state)
 {
   if (domain_state->ephe_info->todo == 0 &&
       domain_state->ephe_info->live == 0 &&
@@ -1829,7 +1829,7 @@ static void handover_ephemerons(caml_domain_state* domain_state)
   CAMLassert (domain_state->ephe_info->todo == 0);
 }
 
-static void handover_finalisers(caml_domain_state* domain_state)
+static void orphan_finalisers(caml_domain_state* domain_state)
 {
   struct caml_final_info* f = domain_state->final_info;
 
@@ -1888,21 +1888,21 @@ static void domain_terminate (void)
        STW sections that has sent an interrupt to this domain. */
 
     caml_finish_marking();
-    handover_ephemerons(domain_state);
-    handover_finalisers(domain_state);
+    orphan_ephemerons(domain_state);
+    orphan_finalisers(domain_state);
 
     /* take the all_domains_lock to try and exit the STW participant set
        without racing with a STW section being triggered */
     caml_plat_lock(&all_domains_lock);
 
     /* The interaction of termination and major GC is quite subtle.
-     *
-     * At the end of the major GC, we decide the number of domains to mark and
-     * sweep for the next cycle. If a STW section has been started, it will
-     * require this domain to participate, which in turn could involve a
-     * major GC cycle. This would then require finish marking and sweeping
-     * again in order to decrement the globals [num_domains_to_mark] and
-     * [num_domains_to_sweep] (see major_gc.c).
+
+       At the end of the major GC, we decide the number of domains to mark and
+       sweep for the next cycle. If a STW section has been started, it will
+       require this domain to participate, which in turn could involve a major
+       GC cycle. This would then require finish marking and sweeping again in
+       order to decrement the globals [num_domains_to_mark] and
+       [num_domains_to_sweep] (see major_gc.c).
      */
 
     if (!caml_incoming_interrupts_queued() &&

--- a/runtime/major_gc.c
+++ b/runtime/major_gc.c
@@ -129,7 +129,7 @@ static atomic_uintnat num_domains_to_final_update_last;
 
    [num_domains_orphaning_finalisers] keeps a count of the number of domains
    currently orphaning finalisers. This counter is only used in the
-   [Phase_sweep_and_mark_main] to determine wheter to proceed to
+   [Phase_sweep_and_mark_main] to determine whether to proceed to
    [Phase_mark_final]. If domains are currently orphaning finalisers, we remain
    in [Phase_sweep_and_mark_main] so that the orphaned finalisers can be
    adopted before moving onto [Phase_mark_final] where the [GC.finalise]
@@ -339,6 +339,10 @@ static void record_ephe_marking_done (uintnat ephe_cycle)
   caml_plat_unlock(&ephe_lock);
 }
 
+/*******************************************************************************
+ * Orphaning and adoption
+ ******************************************************************************/
+
 /* These are biased data structures left over from terminating domains.
 
    Synchronization:
@@ -355,51 +359,6 @@ static struct {
 } orph_structs = {0, NULL};
 
 static caml_plat_mutex orphaned_lock = CAML_PLAT_MUTEX_INITIALIZER;
-
-void caml_add_orphaned_finalisers (struct caml_final_info* f)
-{
-  CAMLassert (caml_gc_phase == Phase_sweep_and_mark_main);
-  CAMLassert (!f->updated_first);
-  CAMLassert (!f->updated_last);
-
-  caml_plat_lock(&orphaned_lock);
-  f->next = orph_structs.final_info;
-  orph_structs.final_info = f;
-  caml_plat_unlock(&orphaned_lock);
-}
-
-/* Called by terminating domain from [handover_finalisers] */
-void caml_final_domain_terminate (caml_domain_state *domain_state)
-{
-  struct caml_final_info *f = domain_state->final_info;
-  if(!f->updated_first) {
-    atomic_fetch_add_verify_ge0(&num_domains_to_final_update_first, -1);
-    f->updated_first = 1;
-  }
-  if(!f->updated_last) {
-    atomic_fetch_add_verify_ge0(&num_domains_to_final_update_last, -1);
-    f->updated_last = 1;
-  }
-}
-
-/* Called by terminating domain from [handover_finalisers] */
-void caml_incr_num_domains_orphaning_finalisers (void)
-{
-  atomic_fetch_add(&num_domains_orphaning_finalisers, +1);
-}
-
-/* Called by terminating domain from [handover_finalisers] */
-void caml_decr_num_domains_orphaning_finalisers (void)
-{
-  atomic_fetch_add_verify_ge0(&num_domains_orphaning_finalisers, -1);
-}
-
-static int no_orphaned_work (void)
-{
-  return
-    atomic_load_acquire(&orph_structs.ephe_list_live) == 0 &&
-    atomic_load_acquire(&orph_structs.final_info) == NULL;
-}
 
 Caml_inline value ephe_list_tail(value e)
 {
@@ -432,8 +391,14 @@ static void orph_ephe_list_verify_status (int status)
 
 static intnat ephe_mark (intnat budget, uintnat for_cycle, int force_alive);
 
-void caml_add_to_orphaned_ephe_list(struct caml_ephe_info* ephe_info)
+void caml_orphan_ephemerons (caml_domain_state* domain_state)
 {
+  struct caml_ephe_info* ephe_info = domain_state->ephe_info;
+  if (ephe_info->todo == 0 &&
+      ephe_info->live == 0 &&
+      ephe_info->must_sweep_ephe == 0)
+    return;
+
   /* Force all ephemerons and their data on todo list to be alive */
   if (ephe_info->todo) {
     while (ephe_info->todo) {
@@ -458,6 +423,57 @@ void caml_add_to_orphaned_ephe_list(struct caml_ephe_info* ephe_info)
     ephe_info->must_sweep_ephe = 0;
     atomic_fetch_add_verify_ge0(&num_domains_to_ephe_sweep, -1);
   }
+  CAMLassert (ephe_info->live == 0);
+  CAMLassert (ephe_info->todo == 0);
+}
+
+void caml_orphan_finalisers (caml_domain_state* domain_state)
+{
+  struct caml_final_info* f = domain_state->final_info;
+
+  if (f->todo_head != NULL || f->first.size != 0 || f->last.size != 0) {
+    /* have some final structures */
+    atomic_fetch_add(&num_domains_orphaning_finalisers, +1);
+    if (caml_gc_phase != Phase_sweep_and_mark_main) {
+      /* Force a major GC cycle to simplify constraints for orphaning
+         finalisers. See note attached to the declaration of
+         [num_domains_orphaning_finalisers] variable in major_gc.c */
+      caml_finish_major_cycle(0);
+    }
+    CAMLassert(caml_gc_phase == Phase_sweep_and_mark_main);
+    CAMLassert (!f->updated_first);
+    CAMLassert (!f->updated_last);
+
+    /* Add the finalisers to [orph_structs] */
+    caml_plat_lock(&orphaned_lock);
+    f->next = orph_structs.final_info;
+    orph_structs.final_info = f;
+    caml_plat_unlock(&orphaned_lock);
+
+    /* Create a dummy final info */
+    domain_state->final_info = caml_alloc_final_info();
+    f = domain_state->final_info;
+    atomic_fetch_add_verify_ge0(&num_domains_orphaning_finalisers, -1);
+  }
+
+  /* [caml_orphan_finalisers] is called in a while loop in [domain_terminate].
+     We take care to decrement the [num_domains_to_final_update*] counters only
+     if we have not already decremented it for the current cycle. */
+  if(!f->updated_first) {
+    atomic_fetch_add_verify_ge0(&num_domains_to_final_update_first, -1);
+    f->updated_first = 1;
+  }
+  if(!f->updated_last) {
+    atomic_fetch_add_verify_ge0(&num_domains_to_final_update_last, -1);
+    f->updated_last = 1;
+  }
+}
+
+static int no_orphaned_work (void)
+{
+  return
+    atomic_load_acquire(&orph_structs.ephe_list_live) == 0 &&
+    atomic_load_acquire(&orph_structs.final_info) == NULL;
 }
 
 static void adopt_orphaned_work (void)
@@ -489,11 +505,13 @@ static void adopt_orphaned_work (void)
   while (f != NULL) {
     myf = domain_state->final_info;
     CAMLassert (caml_gc_phase == Phase_sweep_and_mark_main);
-    CAMLassert (!f->updated_first);
-    CAMLassert (!f->updated_last);
+    /* Since we are in [Phase_sweep_and_mark_main], the current domain has not
+       updated its finalisers. */
     CAMLassert (!myf->updated_first);
     CAMLassert (!myf->updated_last);
+
     if (f->todo_head) {
+      /* Adopt the finalising set. */
       if (myf->todo_tail == NULL) {
         CAMLassert(myf->todo_head == NULL);
         myf->todo_head = f->todo_head;
@@ -503,17 +521,22 @@ static void adopt_orphaned_work (void)
         myf->todo_tail = f->todo_tail;
       }
     }
+
+    /* Adopt the finalisable set */
     if (f->first.young > 0) {
       caml_final_merge_finalisable (&f->first, &myf->first);
     }
     if (f->last.young > 0) {
       caml_final_merge_finalisable (&f->last, &myf->last);
     }
+
     temp = f;
     f = f->next;
     caml_stat_free (temp);
   }
 }
+
+/******************************************************************************/
 
 #define BUFFER_SIZE 64
 

--- a/runtime/major_gc.c
+++ b/runtime/major_gc.c
@@ -80,8 +80,34 @@ uintnat caml_major_cycles_completed = 0;
 static atomic_uintnat num_domains_to_sweep;
 static atomic_uintnat num_domains_to_mark;
 static atomic_uintnat num_domains_to_ephe_sweep;
+
+/* [num_domains_to_final_update_first] and [num_domains_to_final_update_last]
+   are initialised to [num_domains_in_stw] at the start of the cycle. Whenever
+   a domain finishes processing its first or last finalisers, it decrements the
+   appropriate counter.
+
+   Newly created domains increment both the counters. Terminating domain
+   orphans its finalisers and then decrements the counters. See
+   [caml_final_domain_terminate]. */
 static atomic_uintnat num_domains_to_final_update_first;
 static atomic_uintnat num_domains_to_final_update_last;
+
+/* When domains terminate, they will orphan their finalisers. As mentioned in
+   the comment attached to [num_domains_to_final_update_*] counters, a domain
+   will decrement the counters when the corresponding finalisers are processed
+   for that domain. We would like to preserve this invariant when adopting
+   orphaned finalisers. To this end, we orphan and adopt finalisers only in
+   [Phase_sweep_and_mark_main] when [num_domains_to_final_update_*] counters
+   have not been decremented for the domain yet.
+
+   [num_domains_orphaning_finalisers] keeps a count of the number of domains
+   currently orphaning finalisers. This counter is only used in the
+   [Phase_sweep_and_mark_main] to determine wheter to proceed to
+   [Phase_mark_final]. If domains are currently orphaning finalisers, we remain
+   in [Phase_sweep_and_mark_main] so that the orphaned finalisers can be
+   adopted before moving onto [Phase_mark_final] where the [GC.finalise]
+   (finalise first) finalisers are processed. */
+static atomic_uintnat num_domains_orphaning_finalisers = 0;
 
 /* These two counters keep track of how much work the GC is supposed to
    do in order to keep up with allocation. Both are in GC work units.
@@ -316,7 +342,7 @@ void caml_add_orphaned_finalisers (struct caml_final_info* f)
 
 }
 
-/* Called by terminating domain from handover_finalisers */
+/* Called by terminating domain from [handover_finalisers] */
 void caml_final_domain_terminate (caml_domain_state *domain_state)
 {
   struct caml_final_info *f = domain_state->final_info;
@@ -328,6 +354,18 @@ void caml_final_domain_terminate (caml_domain_state *domain_state)
     atomic_fetch_add_verify_ge0(&num_domains_to_final_update_last, -1);
     f->updated_last = 1;
   }
+}
+
+/* Called by terminating domain from [handover_finalisers] */
+void caml_incr_num_domains_orphaning_finalisers (void)
+{
+  atomic_fetch_add(&num_domains_orphaning_finalisers, +1);
+}
+
+/* Called by terminating domain from [handover_finalisers] */
+void caml_decr_num_domains_orphaning_finalisers (void)
+{
+  atomic_fetch_add_verify_ge0(&num_domains_orphaning_finalisers, -1);
 }
 
 static int no_orphaned_work (void)
@@ -1465,6 +1503,8 @@ static int is_complete_phase_sweep_and_mark_main (void)
     atomic_load_acquire (&num_domains_to_sweep) == 0 &&
     atomic_load_acquire (&num_domains_to_mark) == 0 &&
     /* Marking is done */
+    atomic_load_acquire (&num_domains_orphaning_finalisers) == 0 &&
+    /* No domains are orphaning finalisers. */
     atomic_load_acquire(&ephe_cycle_info.num_domains_todo) ==
     atomic_load_acquire(&ephe_cycle_info.num_domains_done) &&
     /* Ephemeron marking is done */

--- a/runtime/major_gc.c
+++ b/runtime/major_gc.c
@@ -73,12 +73,39 @@ struct mark_stack {
 
 uintnat caml_percent_free = Percent_free_def;
 
-/* This variable is only written with the world stopped,
-   so it need not be atomic */
+/* This variable is only written with the world stopped, so it need not be
+   atomic */
 uintnat caml_major_cycles_completed = 0;
 
+/* [num_domains_to_sweep] records the number of domains to sweep in the current
+   major cycle. The number is set to the [num_domains_in_stw] at the start of
+   the cycle and _strictly decreases_ to 0.
+
+   Domains created in a given cycle will not have any sweep work in that cycle.
+   Sweep changes GARBAGE coloured objects in the domain's own pools to FREE
+   (not a distinct colour; object header is set to 0) and adds them to the free
+   list. No object will have the GARBAGE colour in the domain's own pools since
+   the domain starts with an empty pool with no objects and new objects are
+   allocated with colour MARKED. Hence, they do not affect
+   [num_domains_to_sweep].
+
+   Terminating domains terminate after sweeping is complete for their domain.
+   */
 static atomic_uintnat num_domains_to_sweep;
+
+/* [num_domains_to_mark] records the number of domains to mark in the current
+   major cycle. The number is set to the [num_domains_in_stw] at the start of
+   the cycle. The value of [num_domains_to_mark] may decrease or increase.
+
+   [num_domains_to_mark] may grow larger than the value of [num_domains_in_stw]
+   at the start of the cycle. This is because [caml_modify] may push a block
+   into a potentially empty mark stack of the newly spawned domain.
+
+   Terminating domains empty their mark stack before terminating. */
 static atomic_uintnat num_domains_to_mark;
+
+/* [num_domains_to_ephe_sweep] is set to the [participating_count] at the start
+   of the [Phase_sweep_ephe] and strictly decreases. */
 static atomic_uintnat num_domains_to_ephe_sweep;
 
 /* [num_domains_to_final_update_first] and [num_domains_to_final_update_last]
@@ -339,7 +366,6 @@ void caml_add_orphaned_finalisers (struct caml_final_info* f)
   f->next = orph_structs.final_info;
   orph_structs.final_info = f;
   caml_plat_unlock(&orphaned_lock);
-
 }
 
 /* Called by terminating domain from [handover_finalisers] */
@@ -434,7 +460,7 @@ void caml_add_to_orphaned_ephe_list(struct caml_ephe_info* ephe_info)
   }
 }
 
-void caml_adopt_orphaned_work (void)
+static void adopt_orphaned_work (void)
 {
   caml_domain_state* domain_state = Caml_state;
   value orph_ephe_list_live, last;
@@ -460,13 +486,13 @@ void caml_adopt_orphaned_work (void)
     domain_state->ephe_info->live = orph_ephe_list_live;
   }
 
-  myf = domain_state->final_info;
   while (f != NULL) {
+    myf = domain_state->final_info;
+    CAMLassert (caml_gc_phase == Phase_sweep_and_mark_main);
     CAMLassert (!f->updated_first);
     CAMLassert (!f->updated_last);
     CAMLassert (!myf->updated_first);
     CAMLassert (!myf->updated_last);
-    CAMLassert (caml_gc_phase == Phase_sweep_and_mark_main);
     if (f->todo_head) {
       if (myf->todo_tail == NULL) {
         CAMLassert(myf->todo_head == NULL);
@@ -1462,12 +1488,12 @@ static void stw_cycle_all_domains(caml_domain_state* domain, void* args,
   }
 
   /* Ephemerons */
-  // Adopt orphaned work from domains that were spawned and terminated in
-  // the previous cycle.
 #ifdef DEBUG
   orph_ephe_list_verify_status (caml_global_heap_state.UNMARKED);
 #endif
-  caml_adopt_orphaned_work ();
+  /* Adopt orphaned work from domains that were spawned and terminated in the
+     previous cycle. */
+  adopt_orphaned_work ();
   CAMLassert(domain->ephe_info->todo == (value) NULL);
   domain->ephe_info->todo = domain->ephe_info->live;
   domain->ephe_info->live = (value) NULL;
@@ -1499,44 +1525,52 @@ static void stw_cycle_all_domains(caml_domain_state* domain, void* args,
 static int is_complete_phase_sweep_and_mark_main (void)
 {
   return
+    /* Marking is done */
     caml_gc_phase == Phase_sweep_and_mark_main &&
     atomic_load_acquire (&num_domains_to_sweep) == 0 &&
     atomic_load_acquire (&num_domains_to_mark) == 0 &&
-    /* Marking is done */
-    atomic_load_acquire (&num_domains_orphaning_finalisers) == 0 &&
+
     /* No domains are orphaning finalisers. */
+    atomic_load_acquire (&num_domains_orphaning_finalisers) == 0 &&
+
+    /* Ephemeron marking is done */
     atomic_load_acquire(&ephe_cycle_info.num_domains_todo) ==
     atomic_load_acquire(&ephe_cycle_info.num_domains_done) &&
-    /* Ephemeron marking is done */
-    no_orphaned_work();
+
     /* All orphaned ephemerons have been adopted */
+    no_orphaned_work();
 }
 
 static int is_complete_phase_mark_final (void)
 {
   return
+    /* updated finalise first values */
     caml_gc_phase == Phase_mark_final &&
     atomic_load_acquire (&num_domains_to_final_update_first) == 0 &&
-    /* updated finalise first values */
-    atomic_load_acquire (&num_domains_to_mark) == 0 &&
+
     /* Marking is done */
+    atomic_load_acquire (&num_domains_to_mark) == 0 &&
+
+    /* Ephemeron marking is done */
     atomic_load_acquire(&ephe_cycle_info.num_domains_todo) ==
     atomic_load_acquire(&ephe_cycle_info.num_domains_done) &&
-    /* Ephemeron marking is done */
-    no_orphaned_work();
+
     /* All orphaned ephemerons have been adopted */
+    no_orphaned_work();
 }
 
 static int is_complete_phase_sweep_ephe (void)
 {
   return
+    /* All domains have swept their ephemerons */
     caml_gc_phase == Phase_sweep_ephe &&
     atomic_load_acquire (&num_domains_to_ephe_sweep) == 0 &&
-    /* All domains have swept their ephemerons */
-    atomic_load_acquire (&num_domains_to_final_update_last) == 0 &&
+
     /* All domains have updated finalise last values */
-    no_orphaned_work();
+    atomic_load_acquire (&num_domains_to_final_update_last) == 0 &&
+
     /* All orphaned structures have been adopted */
+    no_orphaned_work();
 }
 
 static void stw_try_complete_gc_phase(
@@ -1583,10 +1617,10 @@ static char collection_slice_mode_char(collection_slice_mode mode)
 }
 
 static void major_collection_slice(intnat howmuch,
-                                     int participant_count,
-                                     caml_domain_state** barrier_participants,
-                                     collection_slice_mode mode,
-                                     int force_compaction)
+                                   int participant_count,
+                                   caml_domain_state** barrier_participants,
+                                   collection_slice_mode mode,
+                                   int force_compaction)
 {
   caml_domain_state* domain_state = Caml_state;
   intnat sweep_work = 0, mark_work = 0;
@@ -1676,7 +1710,7 @@ mark_again:
 #ifdef DEBUG
     orph_ephe_list_verify_status (caml_global_heap_state.MARKED);
 #endif
-    caml_adopt_orphaned_work();
+    adopt_orphaned_work();
 
     /* Ephemerons */
     if (caml_gc_phase != Phase_sweep_ephe) {
@@ -1792,10 +1826,10 @@ mark_again:
                                                       - blocks_marked_before));
 
   if (mode != Slice_opportunistic && is_complete_phase_sweep_ephe()) {
+    /* To handle the case where multiple domains try to finish the major cycle
+       simultaneously, we loop until the current cycle has ended, ignoring
+       whether [caml_try_run_on_all_domains] succeeds. */
     saved_major_cycle = caml_major_cycles_completed;
-    /* To handle the case where multiple domains try to finish the major
-      cycle simultaneously, we loop until the current cycle has ended,
-      ignoring whether caml_try_run_on_all_domains succeeds. */
 
     struct cycle_callback_params params;
     params.force_compaction = force_compaction;

